tuta-sdk/rust/sdk/src/services/service_executor.rs

```diff
@@ -0,0 +1,650 @@
+};
+use crate::type_model_provider::TypeModelProvider;
+use crate::{ApiCallError, HeadersProvider};
+use serde::Deserialize;
+use serde::Serialize;
+use std::sync::Arc;
+
+pub struct ServiceExecutor {
+	auth_headers_provider: Arc<HeadersProvider>,
+	crypto_facade: Arc<CryptoFacade>,
+	entity_facade: Arc<dyn EntityFacade>,
+	instance_mapper: Arc<InstanceMapper>,
+	json_serializer: Arc<JsonSerializer>,
+	rest_client: Arc<dyn RestClient>,
+	type_model_provider: Arc<TypeModelProvider>,
+	base_url: String,
+}
+impl ServiceExecutor {
+	pub fn new(
+		auth_headers_provider: Arc<HeadersProvider>,
+		crypto_facade: Arc<CryptoFacade>,
+		entity_facade: Arc<dyn EntityFacade>,
+		instance_mapper: Arc<InstanceMapper>,
+		json_serializer: Arc<JsonSerializer>,
+		rest_client: Arc<dyn RestClient>,
+		type_model_provider: Arc<TypeModelProvider>,
+		base_url: String,
+	) -> Self {
+		Self {
+			auth_headers_provider,
+			crypto_facade,
+			entity_facade,
+			instance_mapper,
+			json_serializer,
+			rest_client,
+			type_model_provider,
+			base_url,
+		}
+	}
+
+	pub async fn get<S>(
+		&self,
+		data: S::Input,
+		params: ExtraServiceParams,
+	) -> Result<S::Output, ApiCallError>
+	where
+		S: GetService,
+	{
+		S::GET(self, data, params).await
+	}
+
+	pub async fn post<S>(
+		&self,
+		data: S::Input,
+		params: ExtraServiceParams,
+	) -> Result<S::Output, ApiCallError>
+	where
+		S: PostService,
+	{
+		S::POST(self, data, params).await
+	}
+
+	pub async fn put<S>(
+		&self,
+		data: S::Input,
+		params: ExtraServiceParams,
+	) -> Result<S::Output, ApiCallError>
+	where
+		S: PutService,
+	{
+		S::PUT(self, data, params).await
+	}
+
+	pub async fn delete<S>(
+		&self,
+		data: S::Input,
+		params: ExtraServiceParams,
+	) -> Result<S::Output, ApiCallError>
+	where
+		S: DeleteService,
+	{
+		S::DELETE(self, data, params).await
+	}
+}
+
+#[async_trait::async_trait]
+impl Executor for ServiceExecutor {
+	async fn do_request<S, I>(
+		&self,
+		data: Option<I>,
+		method: HttpMethod,
+		extra_service_params: ExtraServiceParams,
+	) -> Result<Option<Vec<u8>>, ApiCallError>
+	where
+		S: Service,
+		I: Entity + Serialize + Send,
+	{
+		let url = format!(
+			"{}/rest/{}",
+			if let Some(url) = extra_service_params.base_url {
+				url.clone()
+			} else {
+				self.base_url.clone()
+			},
+			S::PATH,
+		);
+		let model_version: u32 = S::VERSION;
+
+		let body: Option<Vec<u8>> = if let Some(input_entity) = data {
+			let parsed_entity = self
+				.instance_mapper
+				.serialize_entity(input_entity)
+				.map_err(|e| {
+					ApiCallError::internal_with_err(e, "failed to convert to ParsedEntity")
+				})?;
+			let input_type_ref = I::type_ref();
+			let type_model = self
+				.type_model_provider
+				.get_type_model(input_type_ref.app, input_type_ref.type_)
+				.ok_or(ApiCallError::internal(format!(
+					"type {:?} does not exist",
+					input_type_ref
+				)))?;
+			let encrypted_parsed_entity = self.entity_facade.encrypt_and_map_to_literal(
```

Comment:
I think we should not call encrypt for entities that are not encrypted, like I mentioned before